### ISSUE TITLE
Issues with mingw64 building

### ISSUE THREAD

**Issue by iblowmymind** [andrastantos/cray-sim#6]
Hello,
Here are some issues I've found while compiling the emulator using msys2-mingw64:

1) _{bin, lib, obj}/mingw_release not created automatically.

2) Compilation of utils.cpp fails with `error: invalid conversion from 'std::thread::native_handle_type' {aka 'long long unsigned int'} to 'HANDLE' {aka 'void*'} [-fpermissive]`. In sim_lib/utils.cpp, line 48:
Changing line `DWORD_PTR RetVal = SetThreadAffinityMask(aThread.native_handle(), CpuMask);` to `DWORD_PTR RetVal = SetThreadAffinityMask((HANDLE)aThread.native_handle(), CpuMask);` seems to build correctly. (referenced from https://github.com/vsg-dev/VulkanSceneGraph/pull/325).

3) In the linker commands, replace `.a` with `-mt.dll.a` (copy the corresponding "-mt.dll" files from (mingw64 root)/bin directory to _bin/mingw_release)

4) add -I../sim_lib to build commands for utilities (e.g. ds_edit, tap_edit) and append ".exe" to output files

5) Change the PATH argument for *.bat with "mingw" versions of directories.

Will test if the emulator works. Thanks!

**Comment by iblowmymind** [comment]
Also having trouble with networking;
"Can't find appropriate network interface. Please install TAP-windows network driver"
Even though I do have TAP installed (by OpenVPN) and have created a TAP adapter with the specified name in unicos_ramfs.cfg. Is there any way I can get logging info w/out having the emulator's own info screen start?

**Comment by andrastantos** [comment]
Thanks for reporting this. I have not built on Windows in a very long time (in fact, I don't have windows machines anymore), so it's possible that some bit-rot sneaked in. As far as TAP interfaces are considered, I seem to recall that there's something funky about mingw and networking. Any chance you can fire up gdb or similar and try to debug the problem on your side and submit a patch?

January 25, 2023 10:38 AM, "blw" ***@***.*** ***@***.******@***.***>)> wrote:
	Also having trouble with networking;
"Can't find appropriate network interface. Please install TAP-windows network driver"
Even though I do have TAP installed (by OpenVPN) and have created a TAP adapter with the specified name in unicos_ramfs.cfg. Is there any way I can get logging info w/out having the emulator's own info screen start?

**Comment by iblowmymind** [comment]
Trying gdb, but it won't let me display the application in a different tty (because there's only one, /dev/tty, in mingw).

**Comment by andrastantos** [comment]
That's right, I've forgotten about that. What a pain! VS Code is an IDE that seems to work relatively well with gdb, but I haven't tried to pair it with a mingw based build. Lemme see if I can put together a repro.

Andras

January 25, 2023 11:43 PM, "blw" ***@***.*** ***@***.******@***.***>)> wrote:
	Trying gdb, but it won't let me display the application in a different tty (because there's only one, /dev/tty, in mingw).

**Comment by iblowmymind** [comment]
Thanks a ton! I might just try to build it in MSVC instead, but building the entire Boost library from source compared to using a pre-built one in mingw64 seemed daunting... Would I still be able to build with a Boost library package that was pre-built for Windows?

**Comment by iblowmymind** [comment]
MSVC build gets the same error with TAP networking, ugh...

And, pl_extractor & tap_extractor doesn't seem to build (using boost 1.81 after modifying common.props, probably something got deprecated as the rest of the binaries build OK):

It errors out with `'boost::filesystem::basename': Use path::stem() instead` for pl_extractor.cpp and tap_extractor.cpp).
And, cray_sim only builds if the line that says `<Error Condition="!Exists('..\packages\boost.1.72.0.0\build\boost.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\boost.1.72.0.0\build\boost.targets'))" />` in file `cray_sim.vcxproj` is removed. It'll still successfully build.

Also, something I noticed while trying with WSL2 (which does work (even better after I modified the terminal command to use putty.exe), but I can't make it access WSLg nor an Xnest/Xephyr running in the machine, so X11 apps don't work, but anyway):
Your install script for UNICOS seems to assume that:
```
"10.0.1.2")
	CITMKFS_PRELOAD=citmkfs.preload
	CITMNTFS_RELOAD=citmntfs.preload
	UNICOS_MINGLE=unicos.mingle
	UNICOS_PSTLOAD=unicos.pstload
	INSTALLRULES=installrules
;;
```

However, in cray-cd2.iso found in archive.org (which is a 10.0.1.2 install ISO, confirmed by UNICOS_exe\version), these files are labelled like version 10.0.0.2:
```
"10.0.0.2")
	CITMKFS_PRELOAD=CIT.mkfs.preload
	CITMNTFS_RELOAD=CIT.mntfs.preload
	UNICOS_MINGLE=UNICOS.mingle
	UNICOS_PSTLOAD=UNICOS.pstload
	INSTALLRULES=Installrules
;;
```
So, the script fails with an error. I haven't tried modifying it and installing 10.0.1.2 yet, but I will try so.

**Comment by andrastantos** [comment]
Hey!

So, I've managed to repro the TAP driver problem with the simulator and
pushed a fix to the repo (I think what happened was that over the years
OpenVPN changed the way they registred their drivers so the search
routine didn't find them anymore). Would you mind pulling it and see if
it fixes your problems?

Thanks,
Andras

**Comment by iblowmymind** [comment]
I can confirm that it works now, yay!
Had to change to 172.25.0.1 instead of 172.16.0.1 because Windows seems to reserve that for me, but I can ping back and forward, and the network interface seems to work successfully.
Were you able to check on the install script for 10.0.1.2?

**Comment by iblowmymind** [comment]
A suggestion (just an idea) for transferring files that would allow usage on Windows:
Compile something like uFTP (or an incredibly minimal FTP server compatible with UNICOS) for the platform, gzip it and bootstrap it in some way (maybe put it uuencoded in a self-extracting sh script?), send it over, run the FTP server, copy the distr. files (and install script) and close the connection.

**Comment by iblowmymind** [comment]
And also, unicos.generic and param files are not in the .gitignore!

**Comment by iblowmymind** [comment]
```
@@ -161,11 +161,11 @@ case "$VERSION" in
        INSTALLRULES=Installrules
 ;;
 "10.0.1.2")
-       CITMKFS_PRELOAD=citmkfs.preload
-       CITMNTFS_RELOAD=citmntfs.preload
-       UNICOS_MINGLE=unicos.mingle
-       UNICOS_PSTLOAD=unicos.pstload
-       INSTALLRULES=installrules
+       CITMKFS_PRELOAD=CIT.mkfs.preload
+       CITMNTFS_RELOAD=CIT.mntfs.preload
+       UNICOS_MINGLE=UNICOS.mingle
+       UNICOS_PSTLOAD=UNICOS.pstload
+       INSTALLRULES=Installrules
 ;;
 *)
        install_errorexit "Unknown UNICOS version"
```

These changes in the file `install` seems to fix the installation in UNICOS 10.0.1.2, in cray-cd2.iso.

**Comment by andrastantos** [comment]
Thanks for the fixes! I'm going to include them and upload once I get to my home machine.

As far as FTP goes: all what you suggest of course is possible (at least in theory). My goal with this script was to keep as much of the original install process intact as I could. It used rcp, so I'm using rcp. It used rsh/rlogin so I'm using rsh/rlogin. I know, it's a bit overly pedantic. One hope was that this way the script would be less UNICOS version-specific (which it isn't due to Cray having a penchant for renaming key files even between minor releases).

Andras

January 29, 2023 10:53 PM, "blw" ***@***.*** ***@***.******@***.***>)> wrote:
	A suggestion (just an idea) for transferring files that would allow usage on Windows:
Compile something like uFTP (or an incredibly minimal FTP server compatible with UNICOS) for the platform, gzip it and bootstrap it in some way (maybe put it uuencoded in a self-extracting sh script?), send it over, run the FTP server, copy the distr. files (and install script) and close the connection.

**Comment by iblowmymind** [comment]
Thanks!
I just thought that this way, it wouldn’t require one to enable an insecure connection protocol from many years ago. Sure, plain FTP isn’t the most secure thing either, but it’s marginally more modern compared to rsh and has support for Windows. I haven’t been able to find a product that offers an rcp server for Windows yet!
You *could* actually write a very simple server for rcp included in cray-sim as an integrated server that’s cross-platform (support for Windows and Linux), that would be configured to map a certain directory (e.g., maps the ISO mount to /cd and the cray-sim root to /sim, / could be a virtual FS and have nothing except these two directories), and be configured to only listen to the network adapter configured in cray-sim and not allow any outside connections. It could be launched as a background process when cray-sim is ran with unicos_ramfs, e.g, with a configuration parameter that turns the integrated rcp server on and configures which directories will hold the ISO and the emulator’s root directory.
This way, it’ll be more secure compared to having to install an rcp server on the host.
Let me know your thoughts!

**Comment by andrastantos** [comment]
Oh, I see. The thing is, the host is not running any of those servers:
the Cray machine does. You only need the clients for connection. In
that sense, you are not compromising your host machine. I'm not sure if
the client tools are by default packaged with their server components
and if they are enabled by default. That's where the potential security
risk I think really lies.

Andras

**Comment by iblowmymind** [comment]
Without installing rsh-server on the host, it wouldn’t connect and let me copy the files. I think, on the ramfs at least, the rcp on Cray is just a client and we’re running the server as a host.
